Laba №5 Huffman Coding/HuffmanTree.cpp

```diff
@@ -0,0 +1,344 @@
+﻿#include "HuffmanTree.h"
+
+#include "iostream"
+
+
+
+HuffmanTree::Node::Node()
+{}
+
+std::string HuffmanTree::Node::stringifySymbols() const
+{
+    std::string str;
+    for (int i = 0; i < symbols().length(); ++i)
+    {
+        if (symbols().bitValue(i))
+        {
+            str += char(i);
+        }
+    }
+    return str;
+}
+
+int HuffmanTree::Node::frequency() const
+{
+    return m_frequency;
+}
+
+void HuffmanTree::Node::setFrequency(int frequency)
+{
+    m_frequency = frequency;
+}
+
+BoolVector HuffmanTree::Node::symbols() const
+{
+    return m_symbols;
+}
+
+void HuffmanTree::Node::setSymbols(BoolVector symbols)
+{
+    m_symbols = symbols;
+}
+
+HuffmanTree::Node* HuffmanTree::Node::leftChild() const
+{
+    return m_leftChild;
+}
+
+HuffmanTree::Node* HuffmanTree::Node::rightChild() const
+{
+    return m_rightChild;
+}
+
+void HuffmanTree::Node::setLeftChild(Node* leftChild)
+{
+    m_leftChild = leftChild;
+}
+
+void HuffmanTree::Node::setRightChild(Node* rightChild)
+{
+    m_rightChild = rightChild;
+}
+
+void HuffmanTree::Node::printHorizontal(Node* root, int marginLeft, int levelSpacing) const
+{
+    if (root == nullptr) {
+        return;
+    }
+    printHorizontal(root->rightChild(), marginLeft + levelSpacing, levelSpacing);
+    std::cout << std::string(marginLeft, ' ');
+    std::cout << root->frequency() << ' ' << root->stringifySymbols();
+    std::cout << std::endl;
+    printHorizontal(root->leftChild(), marginLeft + levelSpacing, levelSpacing);
+}
+
+HuffmanTree::~HuffmanTree()
+{
+}
+
+void HuffmanTree::clear()
+{
+    clear(m_root);
+    m_root = nullptr;
+}
+
+void HuffmanTree::print()
+{
+    m_root->printHorizontal(m_root, 0, 15);
+}
+
+void HuffmanTree::build(const std::string& textFilename)
+{
+    std::string text;
+
+    std::ifstream textFile;
+    textFile.open(textFilename);
+
+    if (!textFile.is_open())
+        return;
+
+    char ch;
+
+    textFile >> ch;
+
+    while (textFile.good())
+    {
+        text += ch;
+
+        textFile >> ch;
+    }
+
+    textFile.close();
+
+    BoolVector BV(256, 0);
+
+    if (m_root)
+    {
+        clear();
+    }
+
+    std::vector<Node*> nodes;
+    for (const char& symbol : text)
+    {

```

Comment:
1. В случае некорректного файла должны получить пустое дерево.
2. Переменные лучше объявлять непосредственно перед началом использования.
3. Зачем нам переменная `text`, если мы всё равно проходим посимвольно?

```suggestion
void HuffmanTree::build(const std::string& textFilename)
{
    if (m_root)
    {
        clear();
    }

    std::ifstream textFile;
    textFile.open(textFilename);
    char symbol;
    
    if (!textFile.is_open() || !(textFile >> symbol))
        return;

    BoolVector BV(256, 0);
    std::vector<Node*> nodes;
    while (textFile.good())
    {
        // TODO: подсчёт частот
        textFile >> symbol
```